### PR TITLE
Factor out abstract TCPTransport* classes to reduce the netty footprint

### PR DESCRIPTION
**Pull request by s1monw**
Today we have a ton of logic inside the NettyTransport\* codebase. The footprint
of the code that has a direct netty dependency is large and alternative implementations
are pretty hard today since they need to know all about our proticol etc.
This change moves most of the code into TCPTransport\* baseclasses and moves all
the protocol send code together. The base classes now contain the majority of the logic
while NettyTransport\* classes remain to implement the glue code, configuration and optimization.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/transport/TCPTransport.java [hunk 1]
```diff
@@ -0,0 +1,1153 @@
...
+import org.elasticsearch.common.util.concurrent.KeyedLock;
+import org.elasticsearch.indices.breaker.CircuitBreakerService;
+import org.elasticsearch.monitor.jvm.JvmInfo;
+import org.elasticsearch.rest.RestStatus;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.support.TransportStatus;
+import org.jboss.netty.buffer.ChannelBuffer;
+import org.jboss.netty.buffer.ChannelBuffers;
+import org.jboss.netty.channel.ChannelFuture;
+
```
Comment (rmuir) [inline core/src/main/java/org/elasticsearch/transport/TCPTransport.java]: i think these imports are unused. they confused me for a second.

**Comment by rmuir** [comment]
looks good. It reduces NettyTransport to the netty specifics (size cut in half). Personally i do not understand the generics, but this is typical for ES code for me. Maybe we can reduce the generics as a followup.

**Comment by jasontedor** [comment]
Just two comments:

1. This still leaks Netty all over the place, for example `#toChannelBuffer` on the `BytesReference` interface.
2. Bikeshed: Maybe `TCP` -> `Tcp` throughout?

**Comment by s1monw** [comment]
> This still leaks Netty all over the place, for example #toChannelBuffer on the BytesReference interface.

that's right, we also have http but that's not part of this PR. I opened #19129 and #19125 to fix this.

> Bikeshed: Maybe TCP -> Tcp throughout?

sure why not

File: core/src/main/java/org/elasticsearch/transport/TCPTransport.java (hunk 1 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/transport/TCPTransport.java]: yeah you are right... I must have added them back in some refactorings

**Comment by s1monw** [comment]
@jasontedor I merged up with master and fixed all the other leaks, I think you can take another look.. I am renamind the TCP stuff now

**Comment by s1monw** [comment]
I will push this today unless anybody objects

**Comment by jasontedor** [comment]
> I will push this today unless anybody objects

+1

**Pull request merged**